p2p/src/channels/rpc/p2p_channels_rpc_reducer.rs

```diff
@@ -31,35 +58,96 @@ impl P2pChannelsRpcState {
+                    bug_condition!(
+                        "Invalid state for `P2pChannelsRpcAction::RequestSend`, state: {:?}",
+                        rpc_state
+                    );
+                    return Ok(());
                 };
                 *next_local_rpc_id += 1;
                 *local = P2pRpcLocalState::Requested {
                     time: meta.time(),
                     id: *id,
                     request: request.clone(),
                 };
+
+                let dispatcher = state_context.into_dispatcher();
+
+                #[cfg(feature = "p2p-libp2p")]
+                if is_libp2p {
+                    if let Some((query, data)) =
+                        super::libp2p::internal_request_into_libp2p(*request.clone(), *id)
+                    {
+                        dispatcher.push(P2pNetworkRpcAction::OutgoingQuery {
+                            peer_id,
+                            query,
+                            data,
+                        });
+                    }
+                    if let Some(on_init) = on_init {
+                        dispatcher.push_callback(on_init.clone(), (peer_id, *id, *request.clone()));
+                    }
+
+                    return Ok(());
+                }
+
+                dispatcher.push(P2pChannelsRpcEffectfulAction::RequestSend {
+                    peer_id,
+                    id: *id,
+                    request: request.clone(),
+                    on_init: on_init.clone(),
+                });
+                Ok(())
             }
-            P2pChannelsRpcAction::Timeout { .. } => {}
-            P2pChannelsRpcAction::ResponseReceived { .. } => {
-                let Self::Ready { local, .. } = self else {
-                    return;
+            P2pChannelsRpcAction::Timeout { .. } => Ok(()),
+            P2pChannelsRpcAction::ResponseReceived { response, .. } => {
+                let Self::Ready { local, .. } = rpc_state else {
+                    bug_condition!(
+                        "Invalid state for `P2pChannelsRpcAction::ResponseReceived`, state: {:?}",
+                        rpc_state
+                    );
+                    return Ok(());
                 };
                 let P2pRpcLocalState::Requested { id, request, .. } = local else {
-                    return;
+                    bug_condition!(
+                        "Invalid state for `P2pChannelsRpcAction::ResponseReceived`, state: {:?}",
+                        rpc_state
+                    );
+                    return Ok(());
                 };
                 *local = P2pRpcLocalState::Responded {
                     time: meta.time(),
                     id: *id,
                     request: std::mem::take(request),
                 };
+
+                let dispatcher = state_context.into_dispatcher();
+                if let Some(P2pRpcResponse::BestTipWithProof(resp)) = response.as_deref() {
```

Comment:
What other kind of responses can we get and how should be handled?

Reply:
It can be `None` or other variants of `P2pRpcResponse`, they are handle in `node_p2p_effects`